File: 22.Automated-Testing-Unit-Integration-and-E2E/code/typescript/unit_tests_2.ts
export function isStrongPassword(p: string): boolean {
    if (p.length < 8) return false;
    let hasDigit = false;
    let hasUpper = false;
    for (const char of p) {
        if (char >= '0' && char <= '9') hasDigit = true;
        if (char >= 'A' && char <= 'Z') hasUpper = true;
    }
    return hasDigit && hasUpper;
}

describe('isStrongPassword', () => {
    it('rejects short passwords', () => {
        expect(isStrongPassword('short1A')).toBe(false);
    });
    it('accepts valid passwords', () => {
        expect(isStrongPassword('longEnough9')).toBe(true);
    });
});

File: 22.Automated-Testing-Unit-Integration-and-E2E/code/rust/unit_tests_2.rs
pub fn is_strong_password(p: &str) -> bool {
    if p.len() < 8 { return false; }
    let mut has_digit = false;
    let mut has_upper = false;
    for c in p.chars() {
        if c.is_digit(10) { has_digit = true; }
        if c.is_uppercase() { has_upper = true; }
    }
    has_digit && has_upper
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_strong_password() {
        assert!(!is_strong_password("short1A"));
        assert!(is_strong_password("longEnough9"));
    }
}
